pytorch_lightning/utilities/data.py

```diff
@@ -0,0 +1,54 @@
+import torch
+
+from pytorch_lightning.utilities import rank_zero_warn
+
+try:
+    import torch_xla.core.xla_model as xm
+except ImportError:
+    XLA_AVAILABLE = False
+else:
+    XLA_AVAILABLE = True
+
+
+# Utility function to copy data to given device
+# Works for any form of nested lists, tuples or dictionaries containting tensors
+def transfer_data_to_device(batch, device_type, idx=None, warn_on_transfer=False):
+    # Deal with TPUs separately, they don't use device indexes for some reason
+    if device_type == 'tpu' and XLA_AVAILABLE:
+        if callable(getattr(batch, 'to', None)):
+            if warn_on_transfer:
+                rank_zero_warn('Auto transferred data to device {}'.format(xm.xla_device()))
+            return batch.to(xm.xla_device())
+
+    # base case: nothing to do
+    device = torch.device(device_type, idx)
+    if torch.is_tensor(batch) and batch.device == device:
+        return batch
+
+    # object can be directly moved using `cuda` or `to`
+    if callable(getattr(batch, 'cuda', None)) and device_type == 'cuda':
+        if warn_on_transfer:
+            rank_zero_warn('Auto transferred data to device {}'.format(device))
+        return batch.cuda(device=device)
+
+    if callable(getattr(batch, 'to', None)):
+        if warn_on_transfer:
+            rank_zero_warn('Auto transferred data to device {}'.format(device))
```

Comment:
I would move the warning out to the `__call__` , because 1. this utility function is more general (it is used in other parts) and 2. this function is recursive, so if a dict of tensors is passed, the warning would be shown multiple times.

Reply:
I thought about this, a problem with that is I effectively run into the same code duplication problem by trying to detect which device it's on in __call__ since I'd need to recurse into whatever format the data is in again